apps/zotonic_core/src/support/z_auth.erl

```diff
@@ -89,28 +89,48 @@ logon(UserId, Context) ->
 
 %% @doc Allow an admin user to switch to another user account.
 -spec logon_switch( m_rsc:resource_id(), z:context() ) -> {ok, z:context()} | {error, eacces}.
-logon_switch(UserId, Context) ->
-    case m_rsc:exists(UserId, Context) andalso z_acl:is_admin(Context) of
+logon_switch(ToUserId, Context) ->
+    case is_allowed_switch_user(ToUserId, Context) of
         true ->
-            Context1 = z_acl:logon_prefs(UserId, Context),
-            Context2 = z_notifier:foldl(#auth_logon{ id = UserId }, Context1, Context1),
+            Context1 = z_acl:logon_prefs(ToUserId, Context),
+            Context2 = z_notifier:foldl(#auth_logon{ id = ToUserId }, Context1, Context1),
             {ok, Context2};
         false ->
             {error, eacces}
     end.
 
-%% @doc Allow an admin user to switch to another user account.
--spec logon_switch( m_rsc:resource_id(), m_rsc:resource_id(), z:context() ) -> {ok, z:context()} | {error, eacces}.
-logon_switch(UserId, SudoUserId, Context) ->
-    case m_rsc:exists(UserId, Context) andalso z_acl:is_admin( z_acl:logon(SudoUserId, Context) ) of
+%% @doc Allow an admin user to switch to another user account. The ActingUserId is
+%% typically the user_id of the user that initially logged on. This is stored in the
+%% auth_options as 'sudo_user_id' and is set when performing the switch.
+-spec logon_switch(ToUserId, ActingUserId, Context) -> {ok, Context1} | {error, eacces} when
+    ToUserId :: m_rsc:resource_id(),
+    ActingUserId :: m_rsc:resource_id(),
+    Context :: z:context(),
+    Context1 :: z:context().
+logon_switch(ToUserId, ActingUserId, Context) ->
+    OriginalSudoUserContext = z_acl:logon(ActingUserId, Context),
+    case is_allowed_switch_user(ToUserId, OriginalSudoUserContext) of
         true ->
-            Context1 = z_acl:logon_prefs(UserId, Context),
-            Context2 = z_notifier:foldl(#auth_logon{ id = UserId }, Context1, Context1),
+            Context1 = z_acl:logon_prefs(ToUserId, Context),
+            Context2 = z_notifier:foldl(#auth_logon{ id = ToUserId }, Context1, Context1),
             {ok, Context2};
         false ->
             {error, eacces}
     end.
 
+is_allowed_switch_user(?ACL_ADMIN_USER_ID, Context) ->
+    % Hard coded protection against logging in as the admin user from any account that
+    % is not the admin user.
+    z_acl:user(Context) =:= ?ACL_ADMIN_USER_ID;
+is_allowed_switch_user(ToUserId, Context) ->
+    m_rsc:exists(ToUserId, Context)
```

Comment:
During normal logins we also check the user resource is published. Should this also be done here?

Reply:
Good question. My first thought was *yes*, but then I thought that it is actually useful to be able to logon as an user that is not yet published to test their access/account.